romanisim/l3.py

```diff
@@ -17,64 +36,56 @@ def add_objects_to_l3(l3_mos, source_cat, rng=None, seed=None):
 
-    # Generate WCS
-    twcs = wcs.get_mosaic_wcs(l3_mos.meta)
-
-    # Create PSF
-    l3_psf = psf.make_psf(filter_name=filter_name, sca=2, chromatic=False, webbpsf=True)
-
-    # Generate x,y positions for sources
-    coords = np.array([[o.sky_pos.ra.rad, o.sky_pos.dec.rad]
-                       for o in source_cat])
-    sourcecountsall = galsim.ImageF(l3_mos.data.shape[0], l3_mos.data.shape[1], wcs=twcs, xmin=0, ymin=0)
-    xpos, ypos = sourcecountsall.wcs._xy(coords[:, 0], coords[:, 1])
-    xpos_idx = [round(x) for x in xpos]
-    ypos_idx = [round(y) for y in ypos]
-
-    # Create overall scaling factor map
-    Ct_all = (l3_mos.data.value / l3_mos.var_poisson)
-
-    # Cycle over sources and add them to the mosaic
-    for idx, (x, y) in enumerate(zip(xpos_idx, ypos_idx)):
-        # Set scaling factor for injected sources
-        # Flux / sigma_p^2
-        Ct = (l3_mos.data[x][y].value / l3_mos.var_poisson[x][y].value)
-
-        # Create empty postage stamp galsim source image
-        sourcecounts = galsim.ImageF(l3_mos.data.shape[0], l3_mos.data.shape[1], wcs=twcs, xmin=0, ymin=0)
-
-        # Simulate source postage stamp
-        image.add_objects_to_image(sourcecounts, [source_cat[idx]], xpos=[xpos[idx]], ypos=[ypos[idx]],
-                                   psf=l3_psf, flux_to_counts_factor=Ct, bandpass=[filter_name],
-                                   filter_name=filter_name, rng=rng)
-
-        # Scale the source image back by its flux ratios
-        sourcecounts /= Ct
-
-        # Add sources to the original mosaic data array
-        l3_mos.data = (l3_mos.data.value + sourcecounts.array) * l3_mos.data.unit
-
-        # Note for the future - other noise sources (read and flat) need to be implemented
-
-    # Set new poisson variance
-    l3_mos.var_poisson = l3_mos.data.value / Ct_all
+    # Generate WCS (if needed)
+    if wcs is None:
+        wcs = romanisim.wcs.get_mosaic_wcs(l3_mos.meta, shape=l3_mos.data.shape)
+
+    # Create PSF (if needed)
+    if psf is None:
+        psf = romanisim.psf.make_psf(filter_name=filter_name, sca=CENTER_SCA, chromatic=False, webbpsf=True)
+
+    # Create position arrays (if needed)
+    if any(pos is None for pos in [xpos, ypos]):
+        # Create coordinates (if needed)
+        if coords is None:
+            coords = np.array([[o.sky_pos.ra.rad, o.sky_pos.dec.rad]
+                               for o in source_cat])
+            coords_unit = 'rad'
+        # Generate x,y positions for sources
+        sourcecountsall = galsim.ImageF(l3_mos.data.shape[0], l3_mos.data.shape[1], wcs=wcs, xmin=0, ymin=0)
+        xpos, ypos = sourcecountsall.wcs.radecToxy(coords[:, 0], coords[:, 1], coords_unit)
+
+    # Add sources to the original mosaic data array
+    romanisim.image.add_objects_to_image(l3_mos.data, source_cat, xpos=xpos, ypos=ypos,
+                                         psf=psf, flux_to_counts_factor=exptimes, exptimes=exptimes,
```

Comment:
I think the flux_to_counts_factor needs to be multiplied by the AB flux zero point for this filter?  i.e., add_objects_to_image needs to know how many counts to put in, but the fluxes are specified in maggies, so the flux_to_counts_factor needs to know the maggie -> photon relation as well as the exposure time.  You may be doing something related in the photom section of the parameter file, but I didn't see how that connected.

Reply:
You are correct on my using the photom values for that, but I had it outside of add_objects_to_image. 
https://github.com/PaulHuwe/romanisim/blob/RCAL-832_L3Generation/romanisim/l3.py#L361
I can move it in.